include/unifex/bulk.hpp

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) Rishabh Dwivedi <rishabhdwivedi17@gmail.com>
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <unifex/config.hpp>
+#include <unifex/async_trace.hpp>
+#include <unifex/bind_back.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/std_concepts.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/type_list.hpp>
+#include <unifex/type_traits.hpp>
+
+#include <functional>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+namespace _bulk {
+template <typename Receiver, typename Shape, typename Func>
+struct _receiver {
+  struct type;
+};
+
+template <typename Receiver, typename Shape, typename Func>
+using receiver_t = typename _receiver<Receiver, Shape, Func>::type;
+
+template <typename Receiver, typename Shape, typename Func>
+struct _receiver<Receiver, Shape, Func>::type {
+  UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+  UNIFEX_NO_UNIQUE_ADDRESS Shape shape_;
+  UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+
+  template(typename... Values)
+      (requires receiver_of< Receiver, Values...>)
+  void set_value(Values&&... values) && {
+    UNIFEX_TRY {
+      for (Shape i{0}; i < shape_; i++) {
+        func_(i, values...);
+      }
```

Comment:
Here in start of each iteration of loop, there is a chance where we can do cancellation stuff with calling ``unifex::set_done`` and not proceeding for further iteration. ``unifex::bulk_transform`` also chooses this implementation.
```cpp
for (Shape i{0}; i < shape_; i++) {
      if(stop_token.stop_requested()) {
             unifex::set_done(std::move(receiver_));
             return;
      }
}
```
 This looks to be legitimate as the shape_ can be big and time consuming and this can be a good stop point. Is it fine for everyone?

Reply:
That's probably correct. For a small enough granularity of iterations though it would be quite expensive.